rfcs/2021-09-01-8547-accept-metrics-in-datadog-agent-source.md

```diff
@@ -0,0 +1,222 @@
+`dd_url` contains a known pattern (i.e. it has a suffix that matches a Datadog site) some extra hostname manipulation
+happens. But overal, this conditional can be ignore here and if we want unmodified Datadog Agent to use a `dd_url`
+pointing to a Vector deployement, the following route will have to be supported:
+- `/api/v1/validate` for API key validation
+- `/api/v1/check_run` for check submission
+- `/intake/` for events and metadata (possibly others)
+- `/support/flare/` for support flare
+- `/api/v1/series` & `/api/beta/sketches` for metrics submission
+
+Regarding the relative amount of requests:
+- `/api/v1/validate` is requested at a periodic interval (10 seconds).
+- `/intake/` is requested at various (rather long)
+  [intervals](https://github.com/DataDog/datadog-agent/blob/main/pkg/metadata/helper.go#L12-L22).
+- `/support/flare/` is only used by users upon customer support request, but payloads can be rather large (>10MBytes).
+- `/api/v1/series` & `/api/beta/sketches` it accounts for the vast majority of requests, it relays everything received
+  on the Dogstatsd socket by the Agent.
+- `/api/v1/check_run` depends on the number of checks, but it usually accounts for much less data than metrics.
+
+### Implementation
+
+A few details about the Datadog Agents & [Datadog metrics](https://docs.datadoghq.com/metrics/types/):
+- The base structure for all metrics is named
+  [`MetricSample`](https://github.com/DataDog/datadog-agent/blob/main/pkg/metrics/metric_sample.go#L81-L94) and can be
+  of [several types](https://github.com/DataDog/datadog-agent/blob/main/pkg/metrics/metric_sample.go#L20-L31)
+- Major Agent usecases:
+  - Metrics are send from corechecks (i.e. go code)
+    [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/aggregator/sender.go#L227-L252)
+  - Dogstatsd metrics are converted to the `MetricSample` structure
+    [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/dogstatsd/enrich.go#L87-L137) However Datadog Agents
+    metrics are transformed before being sent, ultimately metrics accounts for two differents kind of payload:
+- The count, gauge and rate series kind of payload, sent to `/api/v1/series` using the [JSON schema officially
+  documented](https://docs.datadoghq.com/api/latest/metrics) with few undocumented [additional
+  fields](https://github.com/DataDog/datadog-agent/blob/main/pkg/metrics/series.go#L45-L57), but this align very well
+  with the existing `datadog_metrics` sinks.
+- The sketches kind of payload, sent to `/api/beta/sketches` and serialized as protobuf as shown
+  [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/serializer/serializer.go#L315-L338) (it ultimately lands
+  [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/metrics/sketch_series.go#L103-L269)). Public `.proto`
+  definition can be found
+  [here](https://github.com/DataDog/agent-payload/blob/master/proto/metrics/agent_payload.proto#L47-L81).
```

Comment:
(chatted with Pierre on Slack about this a little bit, but commenting on the RFC to avoid siloing)

My primary concern is that most sketches, including DDSketch, involve some level of configuration -- tunables that influence their memory usage, precision, etc -- and if we're accepting those in a native format, we need a way to know how they were configured.

Said another way, we can't natively merge a DDSketch we receive from the agent with an internal DDSketch unless the alpha/max bins settings match.  We might be able to hard-code it based on what the agent uses, but it could lead to mergeability breaking in the future if the agents changed their settings and Vector was unaware.

I'm going to try and chat with folks internally to understand what our options here are, since if we can't aggregate/merge correctly, then we'll be adding overhead instead of being more efficient.